### PR TITLE
feat(ci): create disk image after a successful full sync test

### LINKED ISSUE #3545: Generate stateful disks containing lightwalletd-supporting zebra-state cache
## Motivation

We currently regenerate disk images with `zebra-state` synced up to Sapling activation for use in our `sync_past_mandatory_checkpoint` test, so that we don't have to spend hours syncing `zebrad` to exercise the relevant code paths of the test. We should do similar for our `lightwalletd` tests, any integration test that requires synced, processed data to be available to a `zebra-state`/`zebrad` instance/test harness, so we don't need to spend extra time syncing.

## Design

Since this is for general testing and not a particular NU activation height, we should aim for something closer to the most recent NU activation on mainnet.

When syncing, ensure that all build features required for lightwalletd support are enabled, and that the config for `zebrad` has all the lightwalletd functionality enabled to populate the state with the tx index, the compactblock cache, etc.

This should run automatically. Currently we regenerate the Sapling checkpoint test state on changes to `zebra-state` format, but that may not be the entire set of changes to trigger on to kick off a new state disk regeneration. 

https://docs.google.com/document/d/1HrDs0eWZs3s8C5rpwCR5-obCO0nNOXtlG5w4jXD9MJM/edit

## Tasks

~- [ ] Create a test similar to `sync_to_mandatory_checkpoint_*` that syncs to a blockheight (NU5 activation?) and stops~ This exists as the full sync test to tip.
- [ ] Create a new / updated GitHub Actions workflow to generate the stateful disk in a similar manner as https://github.com/ZcashFoundation/zebra/blob/main/.github/workflows/regenerate-stateful-test-disks.yml, using the above test, with the appropriate features, config files, etc
- [ ] Make sure the triggers for the workflow fire on changes to state format, lightwalletd support tweaks, etc

### Follow-Up Tasks

- #3511
- https://github.com/ZcashFoundation/zebra/issues/3510

### PR DESCRIPTION
**Pull request by dconnolly**
## Motivation

We need a fully-synced zebrad state (or pretty close) to do fast lightwalletd-zebrad tests (and maybe other things).

## Solution

On successful full sync test, create a disk image in gcloud from the vm image that the test has been writing to, named `zebrad-cache-$COMMIT_SHA_SHORT-mainnet`.

Resolves #3545 

## Review

@gustavovalverde 

### Reviewer Checklist

  - [ ] Works correctly

## Follow Up Work

~It would be nice/handy to include the block height where we stopped syncing in the name of the image, we probably need to extract that with a regex from the sync logs, similar to how we do to get the container name in earlier workflow steps.~ Done!

~We should also include the state version (`v14`, etc) in the name/metadata.~ Done!

### CODE REVIEW

**Comment by gustavovalverde** [comment]
@dconnolly Are we going to add the following to this PR or in a separate one?
- Make sure the triggers for the workflow fire on changes to state format, lightwalletd support tweaks, etc
- Name the image with the block height
- Put the state version in the name

**Comment by teor2345** [comment]
> * Make sure the triggers for the workflow fire on changes to state format, lightwalletd support tweaks, etc

The cached state database version should change whenever we make lightwalletd state changes. So these checks should be the same thing.

(And the same as the existing cached state triggers for the post-Canopy sync workflow.)

**Review by teor2345 — COMMENTED** [review]
If we want to run a testnet sync, we'll need to change:
`FULL_SYNC_MAINNET_TIMEOUT_MINUTES` to:
`FULL_SYNC_TESTNET_TIMEOUT_MINUTES`.

This is a missing feature in the current workflow.

**Comment by dconnolly** [comment]
> @dconnolly Are we going to add the following to this PR or in a separate one?
> 
> * Make sure the triggers for the workflow fire on changes to state format, lightwalletd support tweaks, etc

Oh yes, we have to update the triggers to not just be pre-merge by Mergify, or manual, right? It currently runs every time *.rs changes, deps change, or the workflow yaml changes.

> * Name the image with the block height

We can do some regex to extract this from the stdout as it is now

> * Put the state version in the name

We may have to make Zebra changes to extract this in a nice way

**Comment by teor2345** [comment]
> > * Put the state version in the name
> 
> We may have to make Zebra changes to extract this in a nice way

Zebra logs the state version and network when it opens the state, it looks like:

> Mar 28 19:18:49.565  INFO {zebrad="9e065b5" net="Main"}: zebra_state::service::finalized_state::disk_db: Opened Zebra state cache at /home/user/.cache/zebra/state/v18/mainnet

**Comment by dconnolly** [comment]
> > > * Put the state version in the name
> > 
> > 
> > We may have to make Zebra changes to extract this in a nice way
> 
> Zebra logs the state version and network when it opens the state, it looks like:
> 
> > Mar 28 19:18:49.565  INFO {zebrad="9e065b5" net="Main"}: zebra_state::service::finalized_state::disk_db: Opened Zebra state cache at /home/user/.cache/zebra/state/v18/mainnet

_Nice_ ✨

File: .github/workflows/test-full-sync.yml [hunk 1]
```diff
@@ -212,6 +217,27 @@ jobs:
         env:
           ZEBRA_CONTAINER: ${{ steps.get-container-name.outputs.zebra_container }}
 
+      - name: Get sync height from logs
+        id: get-sync-height
+        run: |
+          INSTANCE_ID=$(gcloud compute instances describe full-sync-${{ env.GITHUB_REF_SLUG_URL }}-${{ env.GITHUB_SHA_SHORT }} --zone ${{ env.ZONE }} --format='value(id)')
+          SYNC_HEIGHT=$(gcloud logging read '(resource.type="gce_instance" AND resource.labels.instance_id="${INSTANCE_ID}")' --format='value(jsonPayload.MESSAGE)' --limit 500 | grep 'finished initial sync to chain tip' |  grep -o -E 'Height\(\d+\)' | grep -o -E '\d+')
```
Comment (teor2345) [inline .github/workflows/test-full-sync.yml]: Question: There is a 500 line limit here. Does this read the earliest lines first, or the latest lines first?

I'm asking for two reasons:
- "finished initial sync to chain tip" comes very close to the end
- "finished initial sync to chain tip" can be logged multiple times

Comment (dconnolly) [reply .github/workflows/test-full-sync.yml]: It reads the newest first by default: 
image
https://cloud.google.com/sdk/gcloud/reference/logging/read#--limit

Comment (dconnolly) [reply .github/workflows/test-full-sync.yml]: I can make this explicit with `--order=desc`

Comment (teor2345) [reply .github/workflows/test-full-sync.yml]: > I can make this explicit with `--order=desc`

Thanks, I think that would be helpful.

**Review by teor2345 — COMMENTED** [review]
Is this blocked by making the full sync faster, so it actually finishes?
Or can we just merge this feature, and make the full sync faster later?

**Comment by dconnolly** [comment]
> Is this blocked by making the full sync faster, so it actually finishes? Or can we just merge this feature, and make the full sync faster later?

We can merge as soon as the test run succeeds, I think everything is fast enough to finish

**Comment by teor2345** [comment]
> > Is this blocked by making the full sync faster, so it actually finishes? Or can we just merge this feature, and make the full sync faster later?
> 
> We can merge as soon as the test run succeeds, I think everything is fast enough to finish

We've only had one successful full sync test in the last few days, the others have timed out:
https://github.com/ZcashFoundation/zebra/actions/workflows/test-full-sync.yml

I don't think that's a blocker for adding cached state. But it is a blocker for using the cached state.

**Comment by teor2345** [comment]
Should we also be running a full sync when the workflow changes?
Or is it ok to just manually test this workflow?

**Comment by dconnolly** [comment]
> https://github.com/ZcashFoundation/zebra/actions/workflows/test-full-sync.yml

They haven't timed out, they finished successfully and ran into gh scripting /gcloud query issues after the sync finished

**Comment by teor2345** [comment]
> > https://github.com/ZcashFoundation/zebra/actions/workflows/test-full-sync.yml
> 
> They haven't timed out, they finished successfully and ran into gh scripting /gcloud query issues after the sync finished

Ah right, that's what I was missing. Thanks for explaining it to me.

We might have some speed issues with the database changes, because all those jobs seem to take slightly longer than 6 hours. But we can fix that next week.

**Review by teor2345 — COMMENTED** [review]
I'm happy to merge this, but I'd like Gustavo to check today's fixes as well.

File: .github/workflows/test-full-sync.yml [hunk 2]
```diff
@@ -79,7 +79,6 @@ jobs:
 
       # Setup Docker Buildx to allow use of docker cache layers from GH
       - name: Set up Docker Buildx
-        id: buildx
```
Comment (gustavovalverde) [inline .github/workflows/test-full-sync.yml]: This is a future proof id needed for arm64 builds. I'd recommend to keep it

File: .github/workflows/test-full-sync.yml [hunk 3]
```diff
@@ -106,7 +105,6 @@ jobs:
 
       # Build and push image to Google Artifact Registry
       - name: Build & push
-        id: docker_build
```
Comment (gustavovalverde) [inline .github/workflows/test-full-sync.yml]: This is a future proof id needed for arm64 builds. I'd recommend to keep it

Comment (gustavovalverde) [inline .github/workflows/test-full-sync.yml]: This is a future proof id needed for arm64 builds. I'd recommend to keep it

File: .github/workflows/test-full-sync.yml [hunk 4]
```diff
@@ -142,6 +140,11 @@ jobs:
         with:
           short-length: 7
 
+      - name: Format network string
+        run: |
+          NETWORK_CAPS=${{ github.event.inputs.network || env.NETWORK }}
+          echo "NETWORK=${NETWORK_CAPS,,}" >> $GITHUB_ENV
+
```
Comment (gustavovalverde) [inline .github/workflows/test-full-sync.yml]:
```suggestion
      - name: Downcase network name for disks
        run: |
          echo LOWER_NET_NAME="${{ github.event.inputs.network || env.NETWORK }}" | awk '{print tolower($0)}' >> $GITHUB_ENV
```

I prefer the way you're doing this. I'll even create a PR to substitute the other implementations available in other YAMLs, as the one available here https://github.com/ZcashFoundation/zebra/blob/main/.github/workflows/test.yml#L368-L370

But if we do it here, we'll have different ways to do it in a single point in time, making it harder for global substitution.

Comment (gustavovalverde) [inline .github/workflows/test-full-sync.yml:147]:
```suggestion
      - name: Downcase network name for disks
        run: |
          echo LOWER_NET_NAME="${{ github.event.inputs.network || env.NETWORK }}" | awk '{print tolower($0)}' >> $GITHUB_ENV
```

I prefer the way you're doing this. I'll even create a PR to substitute the other implementations available in other YAMLs, as the one available here https://github.com/ZcashFoundation/zebra/blob/main/.github/workflows/test.yml#L368-L370

But if we do it here, we'll have different ways to do it in a single point in time, making it harder for global substitution.

File: .github/workflows/test-full-sync.yml [hunk 5]
```diff
@@ -180,41 +181,58 @@ jobs:
...
+          echo "INSTANCE_ID=$INSTANCE_ID" >> $GITHUB_ENV
           echo "Using instance: $INSTANCE_ID"
           while [[ ${CONTAINER_NAME} != *"full-sync-${{ env.GITHUB_REF_SLUG_URL }}-${{ env.GITHUB_SHA_SHORT }}"* ]]; do
               CONTAINER_NAME=$(gcloud logging read 'log_name=projects/${{ env.PROJECT_ID }}/logs/cos_system AND jsonPayload.MESSAGE:full-sync-${{ env.GITHUB_REF_SLUG_URL }}-${{ env.GITHUB_SHA_SHORT }}' --format='value(jsonPayload.MESSAGE)' --limit=1 | grep -o '...-full-sync-${{ env.GITHUB_REF_SLUG_URL }}-${{ env.GITHUB_SHA_SHORT }}-....' | tr -d "'.")
               echo "Using container: ${CONTAINER_NAME} from instance: ${INSTANCE_ID}"
               sleep 10
           done
           CONTAINER_NAME=$(gcloud logging read 'log_name=projects/${{ env.PROJECT_ID }}/logs/cos_system AND jsonPayload.MESSAGE:full-sync-${{ env.GITHUB_REF_SLUG_URL }}-${{ env.GITHUB_SHA_SHORT }}' --format='value(jsonPayload.MESSAGE)' --limit=1 | grep -o '...-full-sync-${{ env.GITHUB_REF_SLUG_URL }}-${{ env.GITHUB_SHA_SHORT }}-....' | tr -d "'.")
-          echo "::set-output name=zebra_container::$CONTAINER_NAME"
+          echo "CONTAINER_NAME=$CONTAINER_NAME" >> $GITHUB_ENV
```
Comment (gustavovalverde) [inline .github/workflows/test-full-sync.yml:200]: @dconnolly do you have any preference on sending the resulting variable value to the global $GITHUB_ENV?

I used GHA's `::set-output` to be able to identify variables being used in different steps, and explicitly define the need of those variables in the steps that need it using `${{ steps.<step>.outputs.<output_id> }}` for traceability.

Comment (gustavovalverde) [inline .github/workflows/test-full-sync.yml:200]: @dconnolly do you have any preference on sending the resulting variable value to the global $GITHUB_ENV?

I used GHA's `::set-output` to be able to identify variables being used in different steps, and explicitly define the need of those variables in the steps that need it, using `${{ steps.<step>.outputs.<output_id> }}` for traceability.

File: .github/workflows/test-full-sync.yml [hunk 6]
```diff
@@ -180,41 +181,58 @@ jobs:
...
           sleep 60
 
-      - name: Full sync mainnet
-        id: full-sync-mainnet
```
Comment (gustavovalverde) [inline .github/workflows/test-full-sync.yml:203]: Would this be a good moment to change the name, or would it be better to do it in a different PR?

Comment (gustavovalverde) [inline .github/workflows/test-full-sync.yml:203]: Would this be a good moment to change the name, or would it be better to do it in a different PR?

File: .github/workflows/test-full-sync.yml [hunk 7]
```diff
@@ -180,41 +181,58 @@ jobs:
...
           --ssh-flag="-o ServerAliveInterval=5" \
-          --command="docker logs --follow ${{ env.ZEBRA_CONTAINER }}"
-        env:
-          ZEBRA_CONTAINER: ${{ steps.get-container-name.outputs.zebra_container }}
+          --command="docker logs --follow ${{ env.CONTAINER_NAME }}"
+
+      - name: Get sync height from logs
+        run: |
+          SYNC_HEIGHT=$(gcloud logging read --format='value(jsonPayload.MESSAGE)' --order="desc" '(resource.labels.instance_id="${{ env.INSTANCE_ID }}" AND jsonPayload.message=~".+finished initial sync to chain tip.+Height\(([0-9]+)\).+")' | grep -oE 'Height\([0-9]+\)' | grep -oE '[0-9]+')
+          echo "SYNC_HEIGHT=$SYNC_HEIGHT" >> $GITHUB_ENV
```
Comment (gustavovalverde) [inline .github/workflows/test-full-sync.yml]: Same comment as above about using `output` variables

Comment (gustavovalverde) [inline .github/workflows/test-full-sync.yml:215]: Same comment as above about using `output` variables

**Review by gustavovalverde — CHANGES_REQUESTED** [review]
I added some comments, most are oriented to keep those changes in a different PR, and around the use of variables.

**Comment by dconnolly** [comment]
Yay

image

File: .github/workflows/test-full-sync.yml (hunk 6 above)
Comment (teor2345) [reply .github/workflows/test-full-sync.yml:203]: Does anything depend on the name right now?

(If we're not using the name in the branch protection rules or mergify rules, go for it!)

**Review by teor2345 — COMMENTED** [review]
I'm happy with this PR, but I don't want to approve it before Gustavo has checked his comments.

**Comment by gustavovalverde** [comment]
@teor2345 @dconnolly I posted a few review comments, are you able to see those? (just to confirm is not happening as last time that I was the only one seeing the comments)

**Comment by teor2345** [comment]
> @teor2345 @dconnolly I posted a few review comments, are you able to see those? (just to confirm is not happening as last time that I was the only one seeing the comments)

I can see your comments, but I figured Deirdre was the best person to fix them.

**Comment by teor2345** [comment]
> Is this blocked by making the full sync faster, so it actually finishes?

Sync speed shouldn't be a blocker for this PR, so we can merge it whenever it's done.

The address index PR series (#3934 to #4038) makes the sync slower by about 2%, which makes the full sync time out more often. But PR #4040 speeds it up by about 7%, so I'm going to merge that PR before all the database changes.

File: .github/workflows/test-full-sync.yml (hunk 2 above)
Comment (dconnolly) [reply .github/workflows/test-full-sync.yml]: I'll undo for now

File: .github/workflows/test-full-sync.yml (hunk 5 above)
Comment (dconnolly) [reply .github/workflows/test-full-sync.yml:200]: Since I needed to use this as an environmental variable in the next step in this job, it made sense to set it as an environmental variable. If the consumer of the variable were some action or other thing that needed them, or another job (vs a follow-on step), I would use output parameters instead

Comment (dconnolly) [reply .github/workflows/test-full-sync.yml:200]: Kinda what this says
https://docs.github.com/en/actions/learn-github-actions/environment-variables#passing-values-between-steps-and-jobs-in-a-workflow

image

Comment (dconnolly) [reply .github/workflows/test-full-sync.yml:200]: Kinda this idea:
https://github.com/ZcashFoundation/zebra/pull/3986/files#r844323388

**Comment by dconnolly** [comment]
@gustavovalverde reverted some of those step id removals, applied your suggested rename, left some comments on the step outputs vs env var questions 🙏

**Review by gustavovalverde — APPROVED** [review]
